### PR TITLE
Support inference of types on deserialization

### PR DESCRIPTION
**Pull request by adsharma**
The API as it stands now needs the caller to explicitly specify the
type to be decoded. Sometimes its desirable to infer that from the
serialized bytes.

```python
from simpleRaft.messages.append_entries import AppendEntriesMessage
from simpleRaft.messages.request_vote import RequestVoteMessage
from simpleRaft.messages.base import BaseMessage
from serde.msgpack import from_msgpack, to_msgpack

ext_dict = {
  BaseMessage.MessageType.AppendEntries: AppendEntriesMessage,
  BaseMessage.MessageType.RequestVote: RequestVoteMessage,
}

a = AppendEntriesMessage("alice", "Bob", 123, 0, 0)
x = to_msgpack(a)
print(x)
b = from_msgpack(BaseMessage, x)
print(b, type(b))  # prints BaseMessage

a = RequestVoteMessage("alice", "Bob", 123, 0, 0)
x = to_msgpack(a, ext_dict=ext_dict)
print(x)
b = from_msgpack(BaseMessage, x, ext_dict=ext_dict)
print(b, type(b))  # prints RequestVoteMessage
a == b  # prints True
```

### CODE REVIEW

**Comment by yukinarit** [comment]
Hi, @adsharma 

Thank you for being interested in pyserde! I understand the concept. 
Can you please write a few more tests? 

Thank you!

**Comment by adsharma** [comment]
Test coverage should look better after I added the tests. Thanks for looking into this!

File: serde/msgpack.py [hunk 1]
```diff
@@ -26,9 +32,17 @@ def deserialize(cls, s, raw=False, use_list=False, **opts):
 
 
 def to_msgpack(obj: Any, se=MsgPackSerializer, **opts) -> bytes:
+    "If `ext_dict` option is specified, `obj` is encoded as a `msgpack.ExtType`"
```
Comment (yukinarit) [inline serde/msgpack.py:35]: Let's make it a docstring!
```suggestion
    """
    If `ext_dict` option is specified, `obj` is encoded as a `msgpack.ExtType`
    """
```

File: serde/msgpack.py [hunk 2]
```diff
@@ -26,9 +32,17 @@ def deserialize(cls, s, raw=False, use_list=False, **opts):
...
     return se.serialize(obj, **opts)
 
 
 def from_msgpack(c: Type[T], s: str, de=MsgPackDeserializer, named=True, **opts) -> Type[T]:
-    fromf = from_dict if named else from_tuple
-    return fromf(c, de.deserialize(s, **opts))
+    "If `ext_dict` option is specified, `c` is ignored and type is inferred from `msgpack.ExtType`"
```
Comment (yukinarit) [inline serde/msgpack.py:40]: Let's make it a docstring!
```suggestion
    """
    If `ext_dict` option is specified, `c` is ignored and type is inferred from `msgpack.ExtType`
    """"
```

**Comment by yukinarit** [comment]
Hi @adsharma!

I left a few comments. Please take a look!

By the way, would you like to change the author of your commit, so that your account is displayed as a contributor of pyserde? 

Screen Shot 2021-01-09 at 20 12 19
 
If you are willing to change, please try the following command to modify the commit author. Thank you!

```sh
git commit --amend --author "adsharma <arun.sharma@gmail.com>"
git push origin master -f
```

File: serde/msgpack.py (hunk 2 above)
Comment (adsharma) [reply serde/msgpack.py]: Ah I thought it was ok to use single quotes for one line docstrings. But you're right pep-257 recommends triple quotes.

**Comment by adsharma** [comment]
> Hi @adsharma!
> 
> I left a few comments. Please take a look!
> 
> By the way, would you like to change the author of your commit, so that your account is displayed as a contributor of pyserde?
> 
> Screen Shot 2021-01-09 at 20 12 19
> 
> If you are willing to change, please try the following command to modify the commit author. Thank you!
> 
> ```shell
> git commit --amend --author <author>
> git push origin master -f
> ```

ok to list me as a contributor, but I use the other email id for all my open source contributions. Updated author per your suggestion.

**Comment by yukinarit** [comment]
@adsharma Thanks, but the author is still not the one linked in Github account? 🤔 
Screen Shot 2021-01-10 at 12 36 09

**Comment by adsharma** [comment]
Probably because the email address was not verified? I've verified it now.
Does it make a difference or do I need to switch the primary email address
as well?

**Comment by yukinarit** [comment]
Screen Shot 2021-01-12 at 7 17 30

LGTM! 👍  Thank you!

**Pull request merged**